### PR TITLE
contracts: Fix some minor bugs around instantiation

### PR DESCRIPTION
**Pull request by athei**
- An error message was wrongly outputted on the node console when instantiating in some cases (using an endowment below the subsistence threshold)
- Break down `NotCallable` into more fine grained an useful error types
- Fixed returning the wrong code size when the initial transfer failed

### CODE REVIEW

File: frame/contracts/src/exec.rs [hunk 1]
```diff
@@ -686,8 +686,11 @@ where
 					contract
 				} else {
 					<ContractInfoOf<T>>::get(&dest)
-						.and_then(|contract| contract.get_alive())
-						.ok_or((Error::<T>::NotCallable.into(), 0))?
+						.ok_or((<Error<T>>::ContractNotFound.into(), 0))
+						.and_then(|contract|
+							contract.get_alive()
+								.ok_or((<Error<T>>::ContractIsTombstone.into(), 0))
+						)?
```
Comment (KiChjang) [inline frame/contracts/src/exec.rs:693]:
```suggestion
						.and_then(ContractInfo::get_alive)
						.ok_or((<Error<T>>::ContractIsTombstone.into(), 0))?
```

File: frame/contracts/src/exec.rs [hunk 2]
```diff
@@ -954,12 +957,23 @@ where
 
 	// The transfer as performed by a call or instantiate.
 	fn initial_transfer(&self) -> DispatchResult {
+		let frame = self.top_frame();
+		let value = frame.value_transferred;
+		let subsistence_threshold = <Contracts<T>>::subsistence_threshold();
+
+		// If the value transferred to a new conteact is less than the subsistence threshold
```
Comment (KiChjang) [inline frame/contracts/src/exec.rs:964]:
```suggestion
		// If the value transferred to a new contract is less than the subsistence threshold
```

File: frame/contracts/src/exec.rs [hunk 3]
```diff
@@ -954,12 +957,23 @@ where
...
+		// we can error out early. This avoids executing the constructor in cases where
+		// we already that the contract has too few balance.
```
Comment (KiChjang) [inline frame/contracts/src/exec.rs:966]:
```suggestion
		// we already know that the contract has too little balance.
```

File: frame/contracts/src/lib.rs [hunk 4]
```diff
@@ -535,9 +535,20 @@ pub mod pallet {
...
+		/// Tombstone cannot be called. Anyone can use `seal_restore_to` in order to revive
+		/// the contract, though.
+		ContractIsTombstone,
+		/// The called contract does not have enough balance to pay for its storage.
+		///
+		/// The contract ran out of balance and is therefore eligible for eviction into a
+		/// tombstone. Anyone can evict the contract by submitting a `claim_surcharge`
+		/// extrinsic. Alternatively, a plain balance transfer can be used in order to
+		/// increase the contracts funds so that it can be called again.
+		RentNotPayed,
```
Comment (KiChjang) [inline frame/contracts/src/lib.rs:551]:
```suggestion
		RentNotPaid,
```

File: frame/contracts/src/exec.rs (hunk 1 above)
Comment (athei) [reply frame/contracts/src/exec.rs:693]: This won't work because you would return an `Option` from `and_then` but it expects a `Result`.

**Review by ascjones — APPROVED** [review]
LGTM

**Review by thiolliere — APPROVED** [review]

File: frame/contracts/src/exec.rs (hunk 1 above)
Comment (KiChjang) [reply frame/contracts/src/exec.rs:693]: My bad, I got myself confused between `Option` and `Result` :D

**Comment by athei** [comment]
bot merge

**Pull request merged**